d 
In this notebook, we will apply Linear Regression to predict the price of a used car in Germany .

Dataset: https://www.kaggle.com/orgesleka/used-cars-database

In [2]:
# File location and type
file_location = "/FileStore/tables/autos.csv"
file_type = "csv"

# CSV options
infer_schema = "false"
first_row_is_header = "true"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
df = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load(file_location)

display(df)

dateCrawled name seller offerType price abtest vehicleType yearOfRegistration gearbox powerPS model kilometer monthOfRegistration fuelType brand notRepairedDamage dateCreated nrOfPictures postalCode lastSeen 2016-03-24 11:52:17 Golf_3_1.6 privat Angebot 480 test null 1993 manuell 0 golf 150000 0 benzin volkswagen null 2016-03-24 00:00:00 0 70435 2016-04-07 03:16:57 2016-03-24 10:58:45 A5_Sportback_2.7_Tdi privat Angebot 18300 test coupe 2011 manuell 190 null 125000 5 diesel audi ja 2016-03-24 00:00:00 0 66954 2016-04-07 01:46:50 2016-03-14 12:52:21 Jeep_Grand_Cherokee_"Overland" privat Angebot 9800 test suv 2004 automatik 163 grand 125000 8 diesel jeep null 2016-03-14 00:00:00 0 90480 2016-04-05 12:47:46 2016-03-17 16:54:04 GOLF_4_1_4__3T�RER privat Angebot 1500 test kleinwagen 2001 manuell 75 golf 150000 6 benzin volkswagen nein 2016-03-17 00:00:00 0 91074 2016-03-17 17:40:17 2016-03-31 17:25:20 Skoda_Fabia_1.4_TDI_PD_Classic privat Angebot 3600 test kleinwagen 2008 manuell 69 fabia 90000 7 diesel skoda nein 2016-03-31 00:00:00 0 60437 2016-04-06 10:17:21 2016-04-04 17:36:23 BMW_316i___e36_Limousine___Bastlerfahrzeug__Export privat Angebot 650 test limousine 1995 manuell 102 3er 150000 10 benzin bmw ja 2016-04-04 00:00:00 0 33775 2016-04-06 19:17:07 2016-04-01 20:48:51 Peugeot_206_CC_110_Platinum privat Angebot 2200 test cabrio 2004 manuell 109 2_reihe 150000 8 benzin peugeot nein 2016-04-01 00:00:00 0 67112 2016-04-05 18:18:39 2016-03-21 18:54:38 VW_Derby_Bj_80__Scheunenfund privat Angebot 0 test limousine 1980 manuell 50 andere 40000 7 benzin volkswagen nein 2016-03-21 00:00:00 0 19348 2016-03-25 16:47:58 2016-04-04 23:42:13 Ford_C___Max_Titanium_1_0_L_EcoBoost privat Angebot 14500 control bus 2014 manuell 125 c_max 30000 8 benzin ford null 2016-04-04 00:00:00 0 94505 2016-04-04 23:42:13 2016-03-17 10:53:50 VW_Golf_4_5_tuerig_zu_verkaufen_mit_Anhaengerkupplung privat Angebot 999 test kleinwagen 1998 manuell 101 golf 150000 0 null volkswagen null 2016-03-17 00:00:00 0 27472 2016-03-31 17:17:06 2016-03-26 19:54:18 Mazda_3_1.6_Sport privat Angebot 2000 control limousine 2004 manuell 105 3_reihe 150000 12 benzin mazda nein 2016-03-26 00:00:00 0 96224 2016-04-06 10:45:34 2016-04-07 10:06:22 Volkswagen_Passat_Variant_2.0_TDI_Comfortline privat Angebot 2799 control kombi 2005 manuell 140 passat 150000 12 diesel volkswagen ja 2016-04-07 00:00:00 0 57290 2016-04-07 10:25:17 2016-03-15 22:49:09 VW_Passat_Facelift_35i__"7Sitzer" privat Angebot 999 control kombi 1995 manuell 115 passat 150000 11 benzin volkswagen null 2016-03-15 00:00:00 0 37269 2016-04-01 13:16:16 2016-03-21 21:37:40 VW_PASSAT_1.9_TDI_131_PS_LEDER privat Angebot 2500 control kombi 2004 manuell 131 passat 150000 2 null volkswagen nein 2016-03-21 00:00:00 0 90762 2016-03-23 02:50:54 2016-03-21 12:57:01 Nissan_Navara_2.5DPF_SE4x4_Klima_Sitzheizg_Bluetooth.Doppelkabine privat Angebot 17999 control suv 2011 manuell 190 navara 70000 3 diesel nissan nein 2016-03-21 00:00:00 0 04177 2016-04-06 07:45:42 2016-03-11 21:39:15 KA_Lufthansa_Edition_450�_VB privat Angebot 450 test kleinwagen 1910 null 0 ka 5000 0 benzin ford null 2016-03-11 00:00:00 0 24148 2016-03-19 08:46:47 2016-04-01 12:46:46 Polo_6n_1_4 privat Angebot 300 test null 2016 null 60 polo 150000 0 benzin volkswagen null 2016-04-01 00:00:00 0 38871 2016-04-01 12:46:46 2016-03-20 10:25:19 Renault_Twingo_1.2_16V_Aut. privat Angebot 1750 control kleinwagen 2004 automatik 75 twingo 150000 2 benzin renault nein 2016-03-20 00:00:00 0 65599 2016-04-06 13:16:07 2016-03-23 15:48:05 Ford_C_MAX_2.0_TDCi_DPF_Titanium privat Angebot 7550 test bus 2007 manuell 136 c_max 150000 6 diesel ford nein 2016-03-23 00:00:00 0 88361 2016-04-05 18:45:11 2016-04-01 22:55:47 Mercedes_Benz_A_160_Classic_Klima privat Angebot 1850 test bus 2004 manuell 102 a_klasse 150000 1 benzin mercedes_benz nein 2016-04-01 00:00:00 0 49565 2016-04-05 22:46:05 2016-04-01 19:56:48 Volkswagen_Scirocco_1.4_TSI_Sport privat Angebot 10400 control coupe 2009 manuell 

In [3]:
df.printSchema()

root
-- dateCrawled: string (nullable = true)
-- name: string (nullable = true)
-- seller: string (nullable = true)
-- offerType: string (nullable = true)
-- price: string (nullable = true)
-- abtest: string (nullable = true)
-- vehicleType: string (nullable = true)
-- yearOfRegistration: string (nullable = true)
-- gearbox: string (nullable = true)
-- powerPS: string (nullable = true)
-- model: string (nullable = true)
-- kilometer: string (nullable = true)
-- monthOfRegistration: string (nullable = true)
-- fuelType: string (nullable = true)
-- brand: string (nullable = true)
-- notRepairedDamage: string (nullable = true)
-- dateCreated: string (nullable = true)
-- nrOfPictures: string (nullable = true)
-- postalCode: string (nullable = true)
-- lastSeen: string (nullable = true)

In [4]:
df.groupby('notRepairedDamage').count().show()

+-----------------+------+
notRepairedDamage| count|
+-----------------+------+
 nein|263390|
 null| 72124|
 ja| 36310|
+-----------------+------+

In [5]:
df.groupby('abtest').count().show()

+-------+------+
 abtest| count|
+-------+------+
control|125543|
 test|135618|
+-------+------+

Both Control and Test are popular so we will keep the 'abtest' column.

In [7]:
(df.count() , len(df.columns))

Out[4]: (371824, 20)

In this dataframe we have 371824 lines and 20 columns

In [9]:
#Explore datatypes

In [10]:
from collections import Counter 
print(Counter(x[1] for x in df.dtypes))

Counter({'string': 20})

In [11]:
df.columns

Out[7]: ['dateCrawled',
 'name',
 'seller',
 'offerType',
 'price',
 'abtest',
 'vehicleType',
 'yearOfRegistration',
 'gearbox',
 'powerPS',
 'model',
 'kilometer',
 'monthOfRegistration',
 'fuelType',
 'brand',
 'notRepairedDamage',
 'dateCreated',
 'nrOfPictures',
 'postalCode',
 'lastSeen']

In [12]:
import pyspark.sql.functions as f

# null values in each column
data_agg = df.agg(*[f.count(f.when(f.isnull(c), c)).alias(c) for c in df.columns])
data_agg.toPandas().T

0 
 
 
 
 
 dateCrawled 
 0 
 
 
 name 
 0 
 
 
 seller 
 1 
 
 
 offerType 
 1 
 
 
 price 
 1 
 
 
 abtest 
 1 
 
 
 vehicleType 
 37900 
 
 
 yearOfRegistration 
 1 
 
 
 gearbox 
 20224 
 
 
 powerPS 
 1 
 
 
 model 
 20499 
 
 
 kilometer 
 1 
 
 
 monthOfRegistration 
 1 
 
 
 fuelType 
 33416 
 
 
 brand 
 1 
 
 
 notRepairedDamage 
 72124 
 
 
 dateCreated 
 1 
 
 
 nrOfPictures 
 1 
 
 
 postalCode 
 1 
 
 
 lastSeen 
 1

In [13]:
df=df.dropna()

Changing Data Types

In [15]:
# casting price target variable to double
new_df = df.withColumn("price", df["price"].cast("double"))
#

In [16]:
new_df = df.withColumn("price", df["price"].cast("double"))
new_df = new_df.withColumn("kilometer",df['kilometer'].cast("double"))
new_df = new_df.withColumn("nrOfPictures",df['nrOfPictures'].cast("double"))
new_df = new_df.withColumn("powerPS",df['powerPS'].cast("int"))

In [17]:
#describe the numeric variables:
new_df.select("price","kilometer","nrOfpictures","powerPS").describe().show()

+-------+-----------------+------------------+------------+------------------+
summary| price| kilometer|nrOfpictures| powerPS|
+-------+-----------------+------------------+------------+------------------+
 count| 261161| 261161| 261161| 261161|
 mean| 8208.17373957061|124058.12889367096| 0.0|126.32613981413763|
 stddev|343174.7890396668| 39848.129258041| 0.0| 145.2308222956912|
 min| 0.0| 5000.0| 0.0| 0|
 max| 9.9999999E7| 150000.0| 0.0| 20000|
+-------+-----------------+------------------+------------+------------------+

Interpretations:

In the price case we have a hight standard deviation comparing to mean , in terms of distribution it indicates that the data are more spread out. 
This is also confirmed by the fact that the minimum value is 0 and the maximum one is 99999999 .
We can remove this range of values by using the quantile.
The number of pictures take only 0 variables so we will remove this column .

In [19]:
new_df=new_df.drop('nrOfpictures')

In [20]:
new_df.dtypes

Out[59]: [('dateCrawled', 'string'),
 ('name', 'string'),
 ('seller', 'string'),
 ('offerType', 'string'),
 ('price', 'int'),
 ('abtest', 'string'),
 ('vehicleType', 'string'),
 ('yearOfRegistration', 'string'),
 ('gearbox', 'string'),
 ('powerPS', 'string'),
 ('model', 'string'),
 ('kilometer', 'double'),
 ('monthOfRegistration', 'string'),
 ('fuelType', 'string'),
 ('brand', 'string'),
 ('notRepairedDamage', 'string'),
 ('dateCreated', 'string'),
 ('nrOfPictures', 'double'),
 ('postalCode', 'string'),
 ('lastSeen', 'string')]

In [21]:
from pyspark.sql.functions import col
new_df.groupby('price').count().sort(col("price")).show()

+-----+-----+
price|count|
+-----+-----+
 0| 3392|
 1| 340|
 2| 4|
 3| 2|
 5| 5|
 9| 4|
 10| 12|
 11| 1|
 13| 3|
 15| 1|
 16| 1|
 20| 4|
 25| 2|
 29| 1|
 30| 4|
 35| 1|
 39| 1|
 40| 3|
 45| 3|
 47| 1|
+-----+-----+
only showing top 20 rows

In [22]:
display(new_df.select('price'))

price 1500.0 3600.0 650.0 2200.0 0.0 2000.0 2799.0 17999.0 1750.0 7550.0 1850.0 10400.0 3699.0 450.0 500.0 2500.0 6900.0 1990.0 690.0 3300.0 18000.0 3500.0 11900.0 1500.0 6990.0 590.0 7999.0 2100.0 13500.0 550.0 39600.0 3000.0 1000.0 7499.0 5500.0 7000.0 14999.0 9700.0 2100.0 800.0 750.0 7999.0 6600.0 7999.0 1400.0 5000.0 20300.0 250.0 600.0 3300.0 1600.0 1800.0 6800.0 2990.0 1.0 14500.0 3850.0 23490.0 1700.0 4290.0 1500.0 22999.0 900.0 3399.0 1300.0 7300.0 8599.0 5900.0 6300.0 3500.0 5950.0 1699.0 4699.0 9290.0 7000.0 1199.0 1750.0 1000.0 1800.0 1100.0 2990.0 2750.0 6990.0 4800.0 14499.0 3450.0 3550.0 8250.0 800.0 41900.0 4200.0 1000.0 800.0 2300.0 4000.0 0.0 9450.0 6890.0 800.0 850.0 8500.0 300.0 1500.0 3200.0 11890.0 2900.0 19750.0 9800.0 1990.0 10500.0 11000.0 2390.0 16449.0 14300.0 5999.0 3700.0 8999.0 25000.0 2790.0 13400.0 4900.0 11299.0 550.0 1400.0 9500.0 1800.0 11000.0 6900.0 1600.0 12999.0 18400.0 999.0 12899.0 10500.0 8500.0 14900.0 500.0 1200.0 1250.0 370.0 350.0 8600.0 2222.0 11800.0 4000.0 1450.0 2800.0 5499.0 5500.0 22290.0 1440.0 1150.0 15700.0 899.0 9200.0 3350.0 2700.0 800.0 9800.0 14700.0 895.0 950.0 6799.0 500.0 11919.0 25900.0 10800.0 650.0 800.0 1690.0 4850.0 4500.0 3900.0 26999.0 750.0 5900.0 56900.0 1100.0 48000.0 8400.0 1.0 800.0 27500.0 14950.0 14900.0 2950.0 12000.0 2500.0 6000.0 8950.0 18700.0 11550.0 2400.0 16150.0 3400.0 3000.0 5500.0 2000.0 890.0 2899.0 4800.0 6799.0 2650.0 1000.0 800.0 5299.0 7150.0 2300.0 5500.0 4400.0 7950.0 16500.0 8100.0 9750.0 7700.0 1195.0 9790.0 1849.0 3300.0 650.0 2600.0 11700.0 600.0 4950.0 1600.0 8499.0 1700.0 5750.0 600.0 8700.0 1200.0 18880.0 11200.0 3900.0 13849.0 13200.0 2500.0 2550.0 3900.0 2450.0 750.0 1799.0 8150.0 3900.0 9999.0 500.0 699.0 999.0 16500.0 6500.0 1490.0 3200.0 3250.0 3400.0 1100.0 1750.0 3300.0 1350.0 8250.0 2500.0 3999.0 21500.0 7000.0 2950.0 1100.0 3600.0 7750.0 6880.0 5600.0 10550.0 17500.0 999.0 2000.0 2700.0 500.0 5555.0 8450.0 650.0 680.0 444.0 4650.0 4450.0 1999.0 9650.0 2200.0 3850.0 15500.0 90500.0 11500.0 2800.0 1490.0 1199.0 350.0 4000.0 6450.0 1690.0 1750.0 2990.0 11000.0 1000.0 650.0 5500.0 1299.0 7600.0 13499.0 1490.0 5500.0 2350.0 6350.0 750.0 1500.0 8500.0 850.0 6700.0 29980.0 2238.0 3350.0 2000.0 14500.0 4199.0 6999.0 4490.0 43900.0 21950.0 8100.0 12650.0 2400.0 24900.0 4300.0 10550.0 200.0 11500.0 20500.0 2990.0 9600.0 1099.0 1700.0 4760.0 16500.0 5300.0 3555.0 19900.0 15200.0 550.0 8700.0 6990.0 1650.0 7900.0 14650.0 3299.0 6500.0 600.0 5800.0 1600.0 8000.0 27500.0 6500.0 2790.0 2390.0 9900.0 1450.0 3600.0 2295.0 950.0 500.0 800.0 1250.0 3250.0 13500.0 4999.0 18300.0 1000.0 36500.0 2700.0 4990.0 6800.0 30900.0 2999.0 1500.0 6500.0 28900.0 1900.0 4900.0 10490.0 28499.0 3999.0 1350.0 600.0 800.0 11500.0 5000.0 2400.0 4800.0 3250.0 11800.0 3500.0 9499.0 3500.0 1999.0 1500.0 400.0 12950.0 2150.0 4999.0 5200.0 8200.0 9099.0 8200.0 2000.0 750.0 1190.0 1200.0 3000.0 1150.0 13980.0 6499.0 8850.0 4000.0 0.0 1990.0 7199.0 1999.0 9990.0 13800.0 2150.0 2500.0 15750.0 750.0 11800.0 2700.0 1650.0 999.0 21500.0 3800.0 11900.0 2100.0 5350.0 6000.0 41900.0 5990.0 0.0 4499.0 7800.0 2900.0 1999.0 1599.0 750.0 1000.0 10500.0 12000.0 1698.0 4500.0 1480.0 8450.0 3700.0 5000.0 21000.0 7500.0 14750.0 2250.0 15500.0 2300.0 12999.0 24800.0 7800.0 6200.0 5300.0 1800.0 2100.0 1700.0 800.0 11850.0 890.0 600.0 750.0 4500.0 4000.0 950.0 5950.0 2650.0 9500.0 9250.0 850.0 2990.0 16000.0 499.0 2500.0 3499.0 1699.0 2400.0 2999.0 13800.0 3450.0 7500.0 11000.0 2100.0 3999.0 13900.0 550.0 250.0 3200.0 11540.0 13199.0 12790.0 600.0 2500.0 13750.0 8200.0 3000.0 2300.0 21999.0 24999.0 2350.0 1750.0 28300.0 14900.0 12800.0 1800.0 5500.0 1800.0 10490.0 37990.0 3800.0 1200.0 10999.0 13900.0 2000.0 5600.0 18000.0 5300.0 13500.0 2850.0 3850.0 6490.0 4550.0 2250.0 7490.0 3250.0 2500.0 699.0 47999.0 9500.0 12990.0 700.0 2750.0 15900.0 13499.0 1300.0 7990.0 4699.0 3500.0 5997.0 2200.0 999.0 1200.0 180.0 999.0 1200.0 5500.0 5250.0 9000.0 4600.0 2500.0 200.0 2799.0 9600.0 1600.

In [23]:
quantile_price = new_df.approxQuantile(['price'], [0.05,0.25, 0.5, 0.75,0.95], 0)
quantile_power=  new_df.approxQuantile(['powerPS'], [0.05,0.25, 0.5, 0.75,0.95], 0)

#price quantile
quantile_05 = quantile_price[0][0]
quantile_25 = quantile_price[0][1]
quantile_50 = quantile_price[0][2]
quantile_75 = quantile_price[0][3]
quantile_95 = quantile_price[0][4]

print('quantile_05: '+str(quantile_05))
print('quantile_25: '+str(quantile_25))
print('quantile_50: '+str(quantile_50))
print('quantile_75: '+str(quantile_75))
print('quantile_95: '+str(quantile_95))

#power quantile
quantile_1 = quantile_power[0][0]
quantile_2 = quantile_power[0][4]

print('quantile_05: '+str(quantile_05))
print('quantile_25: '+str(quantile_25))
print('quantile_50: '+str(quantile_50))
print('quantile_75: '+str(quantile_75))
print('quantile_95: '+str(quantile_95))
print('quantile_1: '+str(quantile_1))
print('quantile_2: '+str(quantile_2))



quantile_05: 500.0
quantile_25: 1500.0
quantile_50: 3850.0
quantile_75: 8600.0
quantile_95: 21900.0
quantile_05: 500.0
quantile_25: 1500.0
quantile_50: 3850.0
quantile_75: 8600.0
quantile_95: 21900.0
quantile_1: 48.0
quantile_2: 239.0

we are going to remove extremely hight prices ( that exceed quantile_95) and very low prices (below quantile_05)

In [25]:
import pyspark.sql.functions as f 
h=new_df.filter((f.col("price")>quantile_05)  & (f.col("price")<quantile_95) & (f.col("powerPS")>quantile_1) & (f.col("powerPS")<quantile_2) )

In [26]:
h.count()

Out[99]: 214192

In [27]:
display(h.select('powerPS'))

powerPS 75 69 102 109 105 140 190 75 136 102 160 231 193 99 50 60 113 218 122 129 70 95 75 177 193 109 61 101 232 60 105 218 160 105 50 156 105 80 82 109 204 75 150 122 87 82 113 86 131 102 109 125 224 101 235 200 178 160 163 77 110 86 120 60 75 54 75 170 131 75 116 170 120 184 109 54 204 75 88 160 163 170 170 69 131 75 64 116 170 150 80 150 131 197 110 120 80 179 105 90 170 102 75 88 131 90 231 163 179 75 140 143 80 125 197 98 218 115 75 75 122 170 109 75 170 102 150 130 75 114 140 211 75 54 90 105 101 118 75 75 140 131 101 184 116 201 75 140 170 235 101 150 95 213 140 102 150 179 90 122 107 69 75 174 125 107 125 220 150 88 140 54 140 170 75 185 54 73 54 77 170 170 90 109 50 150 101 90 75 170 125 116 177 75 143 105 101 54 102 231 90 116 75 192 90 66 115 218 125 60 110 163 54 58 125 140 52 69 115 163 143 131 60 147 80 150 140 105 122 60 115 125 71 95 75 73 97 65 110 82 179 68 136 80 203 150 75 150 82 50 144 120 75 143 160 58 80 143 140 101 101 103 95 75 170 86 140 75 85 60 150 125 90 143 170 105 136 75 98 140 179 109 150 130 140 86 63 170 86 143 81 148 102 75 156 170 109 54 170 125 131 54 90 101 116 129 204 170 122 75 75 69 192 75 97 204 136 140 102 60 105 70 160 122 69 145 60 170 143 136 125 60 150 105 64 180 163 122 193 75 60 101 116 60 231 136 190 115 90 140 116 150 190 75 90 184 122 60 69 75 224 179 156 116 177 197 140 104 116 136 90 107 116 140 140 136 90 58 131 143 136 95 160 147 120 170 131 95 68 58 64 193 122 95 75 193 105 120 60 131 105 75 116 116 170 170 75 102 75 122 88 71 105 65 131 150 105 75 122 120 163 122 125 86 75 110 150 105 140 63 163 102 80 122 101 117 86 193 224 122 95 218 105 141 120 105 75 105 178 90 116 109 190 150 140 156 75 200 109 170 102 113 75 170 115 75 102 193 140 75 150 116 177 101 69 122 170 75 136 75 131 75 132 193 60 165 204 170 155 170 75 163 109 80 75 110 90 234 163 150 129 170 110 224 69 170 122 109 158 150 80 90 116 143 150 60 179 105 197 90 140 218 68 105 140 177 50 150 170 122 102 102 60 117 95 116 69 105 150 130 92 122 135 177 120 65 150 65 204 125 116 220 71 231 143 75 131 95 107 170 105 117 107 230 75 103 90 129 102 82 150 116 131 84 120 143 136 54 200 163 209 136 163 101 170 177 102 90 125 193 105 131 110 146 101 150 150 105 211 90 150 125 150 193 116 179 75 67 60 106 170 69 105 150 180 60 113 163 80 82 101 150 95 160 64 75 107 163 160 90 125 231 75 136 52 235 109 102 177 100 58 120 69 110 75 110 80 72 109 90 75 75 95 150 218 150 75 179 116 218 75 150 65 60 105 73 237 101 105 90 109 71 68 170 200 50 116 179 109 58 101 105 120 140 116 103 101 136 95 101 143 150 170 129 60 80 82 150 54 177 122 102 194 111 105 116 110 150 80 99 160 140 177 194 101 170 101 179 177 163 75 136 105 110 193 145 60 61 50 85 60 140 95 140 87 177 179 136 110 175 96 116 122 105 73 114 116 136 170 54 211 163 122 98 231 160 163 75 86 150 99 116 140 105 90 226 86 231 200 60 115 87 140 55 140 140 204 190 150 68 68 75 82 90 84 75 143 143 86 90 126 140 75 122 75 143 140 150 105 160 232 143 88 192 193 113 131 140 190 131 150 102 170 184 150 60 75 84 125 75 73 95 140 131 122 131 50 200 140 177 116 179 111 163 166 90 75 131 109 204 105 60 170 88 86 122 177 70 122 80 224 150 131 235 116 61 50 118 155 80 193 98 92 163 143 75 140 224 163 116 197 50 170 102 136 155 83 115 163 101 131 177 75 55 75 116 125 125 102 68 128 86 75 75 69 95 102 231 132 136 124 105 110 105 75 150 82 60 219 58 204 90 80 166 90 75 102 218 167 101 71 140 179 75 105 90 69 101 140 129 100 140 131 156 84 126 177 184 116 69 184 192 136 150 122 101 143 90 75 150 174 200 177 150 65 120 116 170 88 131 69 102 88 106 110 170 140 90 71 95 179 54 140 80 120 182 122 75 116 101 131 150 175 101 163 80 179 75 122 116 179 90 125 60 174 122 105 218 155 193

In [28]:
display(h.select('price'))

price 1500.0 3600.0 650.0 2200.0 2000.0 2799.0 17999.0 1750.0 7550.0 1850.0 10400.0 3699.0 2500.0 6900.0 1990.0 690.0 3300.0 18000.0 3500.0 11900.0 1500.0 6990.0 590.0 7999.0 2100.0 13500.0 550.0 3000.0 1000.0 7499.0 5500.0 7000.0 14999.0 9700.0 2100.0 800.0 750.0 7999.0 6600.0 7999.0 1400.0 5000.0 20300.0 600.0 3300.0 1600.0 1800.0 6800.0 2990.0 14500.0 3850.0 1700.0 4290.0 1500.0 900.0 3399.0 1300.0 7300.0 8599.0 5900.0 6300.0 3500.0 5950.0 1699.0 4699.0 9290.0 7000.0 1199.0 1750.0 1000.0 1800.0 1100.0 2990.0 2750.0 6990.0 4800.0 14499.0 3450.0 3550.0 8250.0 800.0 4200.0 1000.0 800.0 2300.0 4000.0 9450.0 6890.0 800.0 850.0 8500.0 1500.0 3200.0 11890.0 2900.0 19750.0 9800.0 1990.0 10500.0 11000.0 2390.0 16449.0 14300.0 5999.0 3700.0 8999.0 2790.0 13400.0 4900.0 11299.0 550.0 1400.0 9500.0 1800.0 11000.0 6900.0 1600.0 12999.0 18400.0 999.0 12899.0 10500.0 8500.0 14900.0 1200.0 1250.0 8600.0 2222.0 11800.0 4000.0 1450.0 2800.0 5499.0 5500.0 1440.0 1150.0 15700.0 899.0 9200.0 3350.0 2700.0 800.0 9800.0 14700.0 895.0 950.0 6799.0 11919.0 10800.0 650.0 800.0 1690.0 4850.0 4500.0 3900.0 750.0 5900.0 1100.0 8400.0 800.0 14950.0 14900.0 2950.0 12000.0 2500.0 6000.0 8950.0 18700.0 11550.0 2400.0 16150.0 3400.0 3000.0 5500.0 2000.0 890.0 2899.0 4800.0 6799.0 2650.0 1000.0 800.0 5299.0 7150.0 2300.0 5500.0 4400.0 7950.0 16500.0 8100.0 9750.0 7700.0 1195.0 9790.0 1849.0 3300.0 650.0 2600.0 11700.0 600.0 4950.0 1600.0 8499.0 1700.0 5750.0 600.0 8700.0 1200.0 18880.0 11200.0 3900.0 13849.0 13200.0 2500.0 2550.0 3900.0 2450.0 750.0 1799.0 8150.0 3900.0 9999.0 699.0 999.0 16500.0 6500.0 1490.0 3200.0 3250.0 3400.0 1100.0 1750.0 3300.0 1350.0 8250.0 2500.0 3999.0 21500.0 7000.0 2950.0 1100.0 3600.0 7750.0 6880.0 5600.0 10550.0 17500.0 999.0 2000.0 2700.0 5555.0 8450.0 650.0 680.0 4650.0 4450.0 1999.0 9650.0 2200.0 3850.0 15500.0 11500.0 2800.0 1490.0 1199.0 4000.0 6450.0 1690.0 1750.0 2990.0 11000.0 1000.0 650.0 5500.0 1299.0 7600.0 13499.0 1490.0 5500.0 2350.0 6350.0 750.0 1500.0 8500.0 850.0 6700.0 2238.0 3350.0 2000.0 14500.0 4199.0 6999.0 4490.0 8100.0 12650.0 2400.0 4300.0 10550.0 11500.0 20500.0 2990.0 9600.0 1099.0 1700.0 4760.0 16500.0 5300.0 3555.0 19900.0 15200.0 550.0 8700.0 6990.0 1650.0 7900.0 14650.0 3299.0 6500.0 600.0 5800.0 1600.0 8000.0 6500.0 2790.0 2390.0 9900.0 1450.0 3600.0 2295.0 950.0 800.0 1250.0 3250.0 13500.0 4999.0 18300.0 1000.0 2700.0 4990.0 6800.0 2999.0 1500.0 6500.0 1900.0 4900.0 10490.0 3999.0 1350.0 600.0 800.0 11500.0 5000.0 2400.0 4800.0 3250.0 11800.0 3500.0 9499.0 3500.0 1999.0 1500.0 12950.0 2150.0 4999.0 5200.0 8200.0 9099.0 8200.0 2000.0 750.0 1190.0 1200.0 3000.0 1150.0 13980.0 6499.0 8850.0 4000.0 1990.0 7199.0 1999.0 9990.0 13800.0 2150.0 2500.0 15750.0 750.0 11800.0 2700.0 1650.0 999.0 21500.0 3800.0 11900.0 2100.0 5350.0 6000.0 5990.0 4499.0 7800.0 2900.0 1999.0 1599.0 750.0 1000.0 10500.0 12000.0 1698.0 4500.0 1480.0 8450.0 3700.0 5000.0 21000.0 7500.0 14750.0 2250.0 15500.0 2300.0 12999.0 7800.0 6200.0 5300.0 1800.0 2100.0 1700.0 800.0 11850.0 890.0 600.0 750.0 4500.0 4000.0 950.0 5950.0 2650.0 9500.0 9250.0 850.0 2990.0 16000.0 2500.0 3499.0 1699.0 2400.0 2999.0 13800.0 3450.0 7500.0 11000.0 2100.0 3999.0 13900.0 550.0 3200.0 11540.0 13199.0 12790.0 600.0 2500.0 13750.0 8200.0 3000.0 2300.0 2350.0 1750.0 14900.0 12800.0 1800.0 5500.0 1800.0 10490.0 3800.0 1200.0 10999.0 13900.0 2000.0 5600.0 18000.0 5300.0 13500.0 2850.0 3850.0 6490.0 4550.0 2250.0 7490.0 3250.0 2500.0 699.0 9500.0 12990.0 700.0 2750.0 15900.0 13499.0 1300.0 7990.0 4699.0 3500.0 5997.0 2200.0 999.0 1200.0 999.0 1200.0 5500.0 5250.0 9000.0 4600.0 2500.0 2799.0 9600.0 1600.0 5999.0 7600.0 7499.0 1100.0 890.0 1099.0 3999.0 4999.0 2799.0 3299.0 3999.0 650.0 8800.0 2000.0 12800.0 2700.0 6999.0 2190.0 7450.0 1850.0 9450.0 2999.0 4800.0 1670.0 1800.0 5299.0 3750.0 4200.0 8380.0 8000.0 2599.0 2650.0 6800.0 2410.0 14900.0 2100.0 3500.0 2250.0 10200.0 2800.0 1500.0 1320.0 7500.0 7000.0 899.0 16500.0 11200.0 21500.0 11500.0 1499.0 14990.

The features we are going to learn are :
- Categorical : 
 - brand
 - fuelType
 - abtest
 - notRepairedDamage
- Numeric:
 - kilometer

# Data Preprocessing

Encode String features

Basic Steps :
- StringIndexer
- OneHotEncoderEstimator
- VectorAssembler

In [33]:
from pyspark.ml.feature import StringIndexer, OneHotEncoderEstimator
from pyspark.ml.feature import VectorAssembler

In [34]:
stage_1 = StringIndexer(inputCol= 'brand', outputCol= 'brand_index')
# define stage 2: transform the column feature_2 to numeric
stage_2 = StringIndexer(inputCol= 'fuelType', outputCol= 'fuelType_index')
# define stage 3: transform the column feature_3 to numeric
stage_3 = StringIndexer(inputCol= 'notRepairedDamage', outputCol= 'notRepairedDamage_index')
# define stage 3: transform the column feature_ to numeric
stage_4= StringIndexer(inputCol= 'abtest', outputCol= 'abtest_index')
# define stage 5: one hot encode the numeric versions of feature 2 and 3 generated from stage 1 and stage 2
stage_5= OneHotEncoderEstimator(inputCols=[stage_1.getOutputCol(),stage_2.getOutputCol(),stage_3.getOutputCol(),stage_4.getOutputCol()], 
                                 outputCols= ['brand_encoded','fuelType_encoded','notRepairedDamage_encoded','abtest_encoded'])
#build a vector that involves all the features
stage_6 =VectorAssembler(inputCols=['kilometer','powerPS']+stage_5.getOutputCols(),outputCol='features')


In [35]:
# setup the pipeline
from pyspark.ml import Pipeline
my_pipeline = Pipeline(stages= [stage_1, stage_2, stage_3, stage_4,stage_5,stage_6])

In [36]:
fitted_model=my_pipeline.fit(h)
transformed_model=fitted_model.transform(h)


In [37]:
data=transformed_model.select('price','features')

Training data

In [39]:
train,test = data.randomSplit([0.7,0.3],seed=12345)

In [40]:
from pyspark.ml.regression import LinearRegression
lr = LinearRegression(featuresCol = 'features', labelCol='price', maxIter=50, regParam=0.3, elasticNetParam=0.8)
lr_model = lr.fit(train)
print("Coefficients: " + str(lr_model.coefficients))
print("Intercept: " + str(lr_model.intercept))

Coefficients: [-0.06852577318222978,48.44578185720891,904.7319270712643,278.0462328109108,-551.8831316829642,153.82814643808172,660.1471025546836,-652.2147404694716,-825.3092858525815,-693.1082477896514,-518.7676641142036,235.02205814622428,944.5801566106804,-544.2828006367852,-97.53665893172571,-716.1569427381995,163.24059133177582,2561.3909406608773,-395.48150629153156,-352.5523308566145,-461.25011102153604,-656.5530810119686,-266.0063918080316,-1093.2732903421918,-174.15416974690064,-1694.2109009236135,-981.8128093134044,-2419.1680987311156,-228.6640336026983,-9.072531193847684,-656.6378648524717,-419.08885207719965,2718.413665618233,-1566.2454825258167,-1542.769516453325,-1069.5349991182486,-1684.090346933683,3140.6175614659423,-1930.5290378855211,-2644.4530596481222,-1291.3997921681173,1202.0365172070642,-1524.3038744966152,161.91486887629395,2675.5941596200937,-1380.5297824505242,2018.3617559018896,-14.243079926609338]
Intercept: 6742.642521230055

In [41]:
trainingSummary = lr_model.summary
print("RMSE: %f" % trainingSummary.rootMeanSquaredError)
print("r2: %f" % trainingSummary.r2)

RMSE: 3056.224807
r2: 0.581515

R squared at 0.58 indicates that in our model, approximate 58% of the variability in 'price' can be explained using the model.

In [43]:
lr_predictions = lr_model.transform(test)
lr_predictions.select("prediction","price","features").show(20)
from pyspark.ml.evaluation import RegressionEvaluator
lr_evaluator = RegressionEvaluator(predictionCol="prediction", \
                 labelCol="price",metricName="r2")



+-------------------+-----+--------------------+
 prediction|price| features|
+-------------------+-----+--------------------+
 2639.8737831873086|525.0|(48,[0,1,20,40,46...|
 3456.353225693666|531.0|(48,[0,1,6,40,46,...|
-225.38111544919957|540.0|(48,[0,1,10,40,47...|
 1699.8563861833272|540.0|(48,[0,1,16,40,46...|
 1423.6866476451469|550.0|(48,[0,1,2,40],[1...|
 517.7595275610674|550.0|(48,[0,1,2,40,46]...|
 517.7595275610674|550.0|(48,[0,1,2,40,46]...|
 1002.2173461331568|550.0|(48,[0,1,2,40,46]...|
 1728.9040739912907|550.0|(48,[0,1,2,40,46]...|
 3264.9260134953665|550.0|(48,[0,1,2,40,46,...|
 -303.7007618372081|550.0|(48,[0,1,2,40,47]...|
 955.8895664502234|550.0|(48,[0,1,2,40,47]...|
 3766.7363818774274|550.0|(48,[0,1,3,40,46]...|
 2396.0114099489683|550.0|(48,[0,1,3,40,46,...|
-1746.0727406648257|550.0|(48,[0,1,4,40],[1...|
 2258.5660713826283|550.0|(48,[0,1,4,40,46]...|
 6860.915347817475|550.0|(48,[0,1,4,40,46]...|
-468.64079254767967|550.0|(48,[0,1,4,40,46,...|
-468.64079254767967|550.0|(48,[0,1,4,40,46,...|
 984.7326631685874|550.0|(48,[0,1,4,40,46,...|
+-------------------+-----+--------------------+
only showing top 20 rows

In [44]:
print("R Squared (R2) on test data = %g" % lr_evaluator.evaluate(lr_predictions))

R Squared (R2) on test data = 0.585602